### PR TITLE
Support for building of docker images for ARM64 platform

### PR DESCRIPTION
**Pull request by stanislav-chlebec**
The changes are inspired by this pull request 
https://github.com/contiv/vpp/commit/535b0614eec1ba8cd0b743d040ead7742c43b14c

There are covered both docker/dev and docker/prod images.

The scripts were rewritten with effort to preserve the previous behaviour when it is possible to pre-set some environment variables then export them and use the newly defined values in scripts instead of defined default values.

Warning: there was removed the sudo command while calling docker command in docker/prod/build.sh what can make problem on not sufficiently set up machines...

### CODE REVIEW

File: docker/dev/build.sh [hunk 1]
```diff
@@ -37,3 +67,23 @@ docker build -f ${DOCKERFILE} \
     --build-arg VERSION=${VERSION} \
     --build-arg COMMIT=${COMMIT} \
     ${DOCKER_BUILD_ARGS} ../..
+
+echo "To push to repository please use command:"
```
Comment (ondrej-fabry) [inline docker/dev/build.sh]: I would prefer not to put info about tag/pushing in build script for the image. That's needed only needed in deployment which we do quite differently.

File: docker/dev/build.sh [hunk 2]
```diff
@@ -1,14 +1,42 @@
...
+# Before run of this script you can set environmental variables
+# IMAGE_TAG, DOCKERFILE, BASE_IMG, GOLANG_OS_ARCH, .. then  export them
+# and to use defined values instead of default ones
 
 cd "$(dirname "$0")"
 
 set -e
 
-IMAGE_TAG=${IMAGE_TAG:-'dev_vpp_agent'}
+#IMAGE_TAG=${IMAGE_TAG:-'dev_vpp_agent'}
```
Comment (ondrej-fabry) [inline docker/dev/build.sh]: Please keep the image-tag name, otherwise this is gonna break build for production image, the name is used as base image there. Let's keep this simple and just use `dev_vpp_agent`, it's local name anyway.

File: docker/dev/build.sh [hunk 3]
```diff
@@ -1,14 +1,42 @@
 #!/bin/bash
+# Before run of this script you can set environmental variables
```
Comment (ondrej-fabry) [inline docker/dev/build.sh]: Actually some prefer not run it this way and only set these variables for single command execution, like this:

```sh
$ BASE_IMG=ubuntu:18.04 ./build.sh
```

File: docker/prod/build.sh [hunk 4]
```diff
@@ -1,9 +1,59 @@
...
+#https://docs.docker.com/registry/spec/manifest-v2-2/#example-manifest-list
 
-sudo docker build  ${DOCKER_BUILD_ARGS} --tag ${IMAGE_TAG} .
+BUILDARCH=`uname -m`
+
+case "$BUILDARCH" in
+  "aarch64" )
+    IMAGE_TAG=${IMAGE_TAG:-'prod_vpp_agent-arm64'}
+    # Dockerfile  for prod_vpp_agent expects that dev-vpp-agent:latest image is built
+    docker tag dev_vpp_agent-arm64:latest dev_vpp_agent:latest
```
Comment (ondrej-fabry) [inline docker/prod/build.sh]: Now I see you are retagging here, but this just unnecessarily complicates overall build process. Always try to minimize parts that don't really help, this just make script longer with extra commands to execute and actually does nothing useful. Keep local image name just `dev_vpp_agent` and `prod_vpp_agent`.

File: docker/dev/build.sh (hunk 1 above)
Comment (stanislav-chlebec) [reply docker/dev/build.sh]: Well I remove that. I will created the push_image.sh instead which directly does pushing.

File: docker/dev/build.sh (hunk 2 above)
Comment (stanislav-chlebec) [reply docker/dev/build.sh]: ok

File: docker/prod/build.sh (hunk 4 above)
Comment (stanislav-chlebec) [reply docker/prod/build.sh]: ok

File: docker/dev/build.sh (hunk 3 above)
Comment (stanislav-chlebec) [reply docker/dev/build.sh]: ok

File: docker/dev/push_image.sh [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    ;;
+
+  "x86_64" )
+    # for AMD64 platform is also used the default image (without suffix -amd64)
+    DEFAULT_IMAGE_NAME='dev-vpp-agent'
+    IMAGE_NAME='dev-vpp-agent-amd64'
+    ;;
+  * )
+    echo "Architecture ${BUILDARCH} is not supported."
+    exit
```
Comment (ondrej-fabry) [inline docker/dev/push_image.sh]: This should most likely exit with 1 so it is obvious that there is some problem.

File: docker/dev/push_image.sh [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+case "$BUILDARCH" in
+  "aarch64" )
+    IMAGE_NAME='dev-vpp-agent-arm64'
+    ;;
+
+  "x86_64" )
+    # for AMD64 platform is also used the default image (without suffix -amd64)
+    DEFAULT_IMAGE_NAME='dev-vpp-agent'
+    IMAGE_NAME='dev-vpp-agent-amd64'
```
Comment (ondrej-fabry) [inline docker/dev/push_image.sh]: You could only store the suffix here and reuse `DEFAULT_IMAGE_NAME`.

File: docker/dev/push_image.sh [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+BUILDARCH=`uname -m`
+
+case "$BUILDARCH" in
+  "aarch64" )
+    IMAGE_NAME='dev-vpp-agent-arm64'
+    ;;
+
+  "x86_64" )
+    # for AMD64 platform is also used the default image (without suffix -amd64)
+    DEFAULT_IMAGE_NAME='dev-vpp-agent'
```
Comment (ondrej-fabry) [inline docker/dev/push_image.sh]: This should most likely be defined just after `REPO_OWNER`.

File: docker/dev/push_image.sh [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+echo "=============================="
+echo "Architecture: ${BUILDARCH}"
+echo "=============================="
+
+case "${BRANCH_NAME}" in
+  "master" )
+    if [ ${BRANCH_HEAD_TAG} != "undefined" ] ; then
+      if [ ${BUILDARCH} = "x86_64" ] ; then
+        # for AMD64 platform is used also the default image (without suffix -amd64)
+        docker tag ${LOCAL_IMAGE}:latest ${REPO_OWNER}/${DEFAULT_IMAGE_NAME}:${BRANCH_HEAD_TAG}
```
[the remark this answers is not in the document]
Comment (stanislav-chlebec) [reply docker/dev/push_image.sh]: Done.

File: docker/dev/push_image.sh (hunk 5 above)
Comment (stanislav-chlebec) [reply docker/dev/push_image.sh]: Done.

File: docker/dev/push_image.sh (hunk 6 above)
Comment (stanislav-chlebec) [reply docker/dev/push_image.sh]: Done.

File: docker/dev/push_image.sh (hunk 7 above)
Comment (stanislav-chlebec) [reply docker/dev/push_image.sh]: Done.

File: docker/dev/push_image.sh [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+# detect branch name
+BRANCH_NAME="$(git symbolic-ref HEAD 2>/dev/null)" || BRANCH_NAME="(unnamed branch)"     # detached HEAD
+BRANCH_NAME=${BRANCH_NAME##refs/heads/}
+BRANCH_HEAD_TAG=${BRANCH_HEAD_TAG:-"`git name-rev --name-only --tags HEAD`"}
+VERSION=$(git describe --always --tags --dirty)
+
+LOCAL_IMAGE=${LOCAL_IMAGE:-'dev_vpp_agent:latest'}
+REPO_OWNER=${REPO_OWNER:-'ligato'}
+DEFAULT_IMAGE_NAME=${DEFAULT_IMAGE_NAME:-'dev-vpp-agent'}
```
Comment (ondrej-fabry) [inline docker/dev/push_image.sh]: And as a final thing, just shorten this to `IMAGE_NAME` since we don't have that variable now and I'll merge it.

Comment (stanislav-chlebec) [reply docker/dev/push_image.sh]: Done:-)
Now it is almost perfect.

File: docker/prod/push_image.sh [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+# Usage: examples
+#    ./push_image.sh
+#    BRANCH_HEAD_TAG='git describe' ./push_image.sh
+#    REPO_OWNER=stanislavchlebec BRANCH_HEAD_TAG=`git describe` ./push_image.sh
+#    LOCAL_IMAGE='prod_vpp_agent:latest' DEFAULT_IMAGE_NAME='vpp-agent' ./push_image.sh
+
+# Warning: use only IMMEDIATELY after docker/dev/build.sh to prevent INCONSISTENCIES such as 
+#          a) after building image you switch to other branch which will result in mismatch of version of image and its tag
+#          b) you do not build the new image but only simply run this script which will result in mismatch version of image and its tag because the image is older than repository 
+LOCAL_IMAGE='prod_vpp_agent:latest' DEFAULT_IMAGE_NAME='vpp-agent' ../dev/push_image.sh
```
Comment (ondrej-fabry) [inline docker/prod/push_image.sh]: Forgot to change here. Perhaps the push script should be only one and located in `docker` directory directly.

Comment (stanislav-chlebec) [reply docker/prod/push_image.sh]: I would prefer to leave it like it is.
The user then does not have to write variables before script
so IMO it has sense to make it like it is.

Comment (stanislav-chlebec) [reply docker/prod/push_image.sh]: The next thing is that how it is described in the beginning of the script there is close connection to the build.sh script - they IMHO should be one next to other one.

Comment (ondrej-fabry) [reply docker/prod/push_image.sh]: Okay then. `DEFAULT_IMAGE_NAME` needs to be fixed anyway though.

Comment (stanislav-chlebec) [reply docker/prod/push_image.sh]: Yes - that is what you refered that I forgot:-) I did not realised.
I do it.
